Delegator 1-to-1 switch out

rclough opened the pull request:
This commit implements previously discussed backend delegator functionality, by first swapping out the existing commit functions, and ensuring it passes all historical tests.

As debugging continues on JIJ, the delegator will be used to take care of special cases, preventing commits from runtime objects, and also allowing easier and more central future controls over write access to dagobah.

rclough commented:
This should be sufficient for review. I'm investigating why travis is failing cause all tests pass on my machine

rclough commented:
it appears that running nosetests twice in a row generates a duplicate dagobah id. This does not happen while running dagobahd manually multiple times, investigating

rclough commented:
confusingly, these errors ar completely unlike the travis errors which I haven't the slighted slue how theyre happening. Any idea what ight be going on @thieman ?

thieman commented:
It looks like from the logs that you're doing SQLite stuff as part of these tests. The test previously was trying pretty hard to use the BaseBackend: https://github.com/thieman/dagobah/blob/master/tests/test_api.py#L17

You probably need to change that setup code to muck with your new delegator. I'll review the rest of this on Thursday probably.

rclough commented:
Ahhhh I thought I had removed all the SQLite stuff a while ago.

rclough commented:
Turns out something is wrong with setup/teardown, which is why its trying to default to sqlite.

thieman commented:
The rest of the SQLite code is alive and well in the job-in-job branch, you might want to do some rebasing to get caught up with the relevant changes in master.

Your immediate problem remains in the setup of test_api here https://github.com/thieman/dagobah/blob/dev-job-in-job/tests/test_api.py#L28

The `set_backend` method on the Dagobah class wasn't upgraded to propagate the change to the Delegator, so it's still using the backend that was given when the Dagobah object was initialized (which is SQLite). As a design note, the `set_backend` method and any public exposure of the `backend` property at all should no longer exist on the Dagobah and should instead live on the Delegator. So you need to replace the call to `set_backend` in the test setup with a refactored call that instead goes to the Delegator itself.

rclough commented:
Yeah, I wanted to remove any reference to backend from the classes, but I decided only to restrict *writes* after taking a look at the code because I remember it would have been a pretty hairy issue to remove read behavior in some cases. It's been a few weeks since I looked at that code though otherwise I'd be more specific. In lieu of that I would love to be able to make the backend classes of that capable "read-only". 

For now I'm gonna update with a fix for this issue that passes the correct backend to the delegator

rclough commented:
just searching briefly for backend read references in the code, I think the issue here is that the delegator was largely to delegate commit behavior, in order to support read behavior, it would either have to basically proxy-implement a bunch of functions from the backend (decode_import_json, get_dagobah_backend, get_new_job_id, delete_dagobah, etc), with no behavior change since it's just a read. Which I think is why I thought a solution of making the backend read-only (and giving the delegator a writable backend) makes more sense

Anyways, tests pass now \o/

will attempt a rebase soon, should be interesting considering the core.py split up

rclough commented:
Or rather I will do a rebase after this PR

rclough commented:
Any thoughts or opposition to merging?

thieman commented on dagobah/core/delegator.py:
```diff
@@ -0,0 +1,41 @@
+    processing.
+    """
+
+    def __init__(self, backend=BaseBackend()):
```
don't think you want to provide a default backend here, we already do this in the Dagobah class constructor (can't really remember why), but failing to provide one to this class is probably a programming error

rclough replied:
Good point, I think I modeled this off of the existing default behavior you mentioned, but should probably be no default.

thieman commented on dagobah/core/delegator.py:
```diff
@@ -0,0 +1,41 @@
+        logger.debug('Committing job {0}'.format(job.name))
+        self.backend.commit_job(job._serialize())
+        self.commit_dagobah(job.parent)
+        return
```
extraneous return

thieman commented on dagobah/core/delegator.py:
```diff
@@ -0,0 +1,41 @@
+                     format(cascade))
+        self.backend.commit_dagobah(dagobah._serialize())
+        if cascade:
+            [self.commit_job(job) for job in dagobah.jobs]
```
doing a lot of extraneous I/O here, maybe refactor the cascade case to only commit the dagobah at the end of all the job commits?

rclough replied:
yeah, this is old code that I just ported over to the class (see dagobah.py L122 in the changes, being deleted by this PR). Dagobah has always been a bit excessive. Might be a good time for refactoring though.

rclough replied:
I'm curious as to why this is done the way it's done actually, you're right that it is extremely heavy on IO. I feel like an op like this should really be wrapped in a transaction, as committing a bunch of jobs and then dagobah could theoretically leave the DB in an invalid state if it breaks somewhere in between or something else crashes.

In the meantime, I'm going to try to disable committing dagobah in the cascade case and move the final dagobah commit to the end and see if it passes tests/looks okay when actually used

thieman commented on dagobah/core/job.py:
```diff
@@ -54,7 +55,8 @@ def __init__(self, parent, backend, job_id, name):
 
-        self.commit()
+        self.delegator = CommitDelegator(backend)
+        self.delegator.commit_job(self)
```
Shouldn't this use the delegator instance from the parent?

thieman commented on dagobah/core/jobtask.py:
```diff
@@ -15,6 +16,9 @@ def __init__(self, parent_job, target_job_name, task_name):
         self.target_job_name = target_job_name
         self.name = task_name
 
+        self.delegator = CommitDelegator(self.parent_job.backend)
```
Same comment on using the parent Dagobah's delegator

thieman commented on dagobah/core/task.py:
```diff
@@ -31,6 +32,9 @@ def __init__(self, parent_job, command, name,
         self.name = name
         self.hostname = hostname
 
+        self.delegator = CommitDelegator(self.backend)
```
parent again

thieman commented:
Had a few more comment, design seems fine. Once you get this merged probably worth cleaning up the SQLite stuff.

rclough commented:
With you on those parent delegator things, I gotta run but I'll do that first thing in the AM. 

I tried doing a rebase today on the branch as a test run but defintiely gonna rebase (which will remove sqlite stuff) as soon as merge is done

rclough commented:
refactored that, added an option to not commit dagobah when committing the job.

Also I rescind comments about transactions in light of me remembering mongo doesn't support them.

rclough commented:
Everything in the last review should be taken care of now, will merge monday if no further comment
